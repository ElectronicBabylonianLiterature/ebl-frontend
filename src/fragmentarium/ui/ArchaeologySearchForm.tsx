import React from 'react'
import withData from 'http/withData'
import Select from 'react-select'
import FragmentService from 'fragmentarium/application/FragmentService'
import './ArchaeologySearchForm.sass'

export default withData<
  {
    onChange: (value: string | null) => void
    value?: string | null
  },
  { fragmentService: FragmentService },
  ReadonlyArray<ReadonlyArray<string>>
>(
  ({ data, value, onChange }) => {
    const options = data.map((archaeology) => ({
      value: archaeology[0],
      label: archaeology[0],
    }))
    const defaultOption = value ? { value: value, label: value } : null

    return (
      <Select
        aria-label="select-archaeology"
        placeholder="Provenance"
        options={options}
        value={defaultOption}
        onChange={(selection) => {
          onChange(selection?.value || null)
        }}
        isSearchable={true}
        classNamePrefix={'archaeology-selector'}
        isClearable
      />
    )
  },
  (props) => props.fragmentService.fetchArchaeologies()
)
