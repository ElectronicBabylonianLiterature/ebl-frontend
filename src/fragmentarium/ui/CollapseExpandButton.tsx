import React, { useState } from 'react'
import Button from 'react-bootstrap/Button'

type CollapseExpandButtonProps = {
  onToggle
  initialCollapsed?: boolean
}

const CollapseExpandButton: React.FC<CollapseExpandButtonProps> = ({
  onToggle,
  initialCollapsed = false,
}) => {
  const [isCollapsed, setIsCollapsed] = useState(initialCollapsed)

  const handleClick = () => {
    const newCollapsedState = !isCollapsed
    setIsCollapsed(newCollapsedState)
    onToggle(newCollapsedState)
  }

  return (
    <Button variant="primary" onClick={handleClick}>
      {isCollapsed ? 'Show' : 'Hide'} Image Column
    </Button>
  )
}

export default CollapseExpandButton
