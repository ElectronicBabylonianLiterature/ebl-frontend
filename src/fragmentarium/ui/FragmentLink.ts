import React from 'react'
import { Link } from 'react-router-dom'
import queryString from 'query-string'

export function createFragmentUrl(number) {
  // Double encoding is needed due to https://github.com/ReactTraining/history/issues/505
  return `/fragmentarium/${encodeURIComponent(encodeURIComponent(number))}`
}

export function createFragmentUrlWithFolio(number, folio) {
  const query = queryString.stringify(
    { tab: 'folio', folioName: folio.name, folioNumber: folio.number },
    { strict: false }
  )
  return `${createFragmentUrl(number)}?${query}`
}

export function createFragmentUrlWithTab(number, tab) {
  const query = queryString.stringify({ tab }, { strict: false })
  return `${createFragmentUrl(number)}?${query}`
}

export default function FragmentLink({ number, children, folio, ...props }) {
  const link = folio
    ? createFragmentUrlWithFolio(number, folio)
    : createFragmentUrl(number)
  return (
    <Link to={link} {...props}>
      {children}
    </Link>
  )
}
