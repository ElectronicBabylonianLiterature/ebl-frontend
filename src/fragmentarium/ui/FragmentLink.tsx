import React, { ReactNode } from 'react'
import { Link } from 'react-router-dom'
import { stringify } from 'query-string'
import Folio from 'fragmentarium/domain/Folio'

export function createFragmentUrl(number) {
  // Double encoding is needed due to https://github.com/ReactTraining/history/issues/505
  return `/fragmentarium/${encodeURIComponent(encodeURIComponent(number))}`
}

export function createFragmentUrlWithFolio(number, folio) {
  const query = stringify(
    { tab: 'folio', folioName: folio.name, folioNumber: folio.number },
    { strict: false }
  )
  return `${createFragmentUrl(number)}?${query}`
}

export function createFragmentUrlWithTab(number, tab) {
  const query = stringify({ tab }, { strict: false })
  return `${createFragmentUrl(number)}?${query}`
}

export default function FragmentLink({
  number,
  children,
  folio,
  ...props
}: {
  number: string
  children: ReactNode
  folio: Folio | null
}) {
  const link = folio
    ? createFragmentUrlWithFolio(number, folio)
    : createFragmentUrl(number)
  return (
    <Link to={link} {...props}>
      {children}
    </Link>
  )
}
FragmentLink.defaultProps = { folio: null }
