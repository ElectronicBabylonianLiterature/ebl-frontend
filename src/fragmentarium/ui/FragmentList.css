.FragmentList__Description {
  width: 40%;
}
.FragmentList__References.Id {
  padding-left: 4%;
}
