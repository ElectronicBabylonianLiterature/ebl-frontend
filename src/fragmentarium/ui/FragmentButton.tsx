import React, { useState } from 'react'
import { Button } from 'react-bootstrap'
import { withRouter, RouteComponentProps } from 'react-router-dom'
import Promise from 'bluebird'
import ErrorAlert from 'common/ErrorAlert'
import Spinner from 'common/Spinner'
import { createFragmentUrl } from './FragmentLink'
import usePromiseEffect from 'common/usePromiseEffect'
import { FragmentInfo } from 'fragmentarium/domain/fragment'

type Props = {
  query: () => Promise<FragmentInfo>
  children?: React.ReactNode
} & RouteComponentProps

function FragmentButton({
  query,
  history,
  children
}: Props) {
  const [error, setError] = useState(null)
  const [isLoading, setIsLoading] = useState(false)
  const [setPromise, cancelPromise] = usePromiseEffect()

  const onError = error => {
    setIsLoading(false)
    setError(error)
  }

  const navigateToFragment = fragmentInfo =>
    history.push(createFragmentUrl(fragmentInfo.number))

  const handleClick = event => {
    cancelPromise()
    setIsLoading(true)
    setError(null)
    setPromise(
      query()
        .then(navigateToFragment)
        .catch(onError)
    )
  }

  return (
    <>
      <Button variant="primary" onClick={handleClick}>
        {isLoading ? <Spinner /> : children}
      </Button>
      <ErrorAlert error={error} />
    </>
  )
}

export default withRouter(FragmentButton)
