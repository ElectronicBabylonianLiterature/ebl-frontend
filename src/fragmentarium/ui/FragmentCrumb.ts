import { Crumb } from 'common/Breadcrumbs'
import { createFragmentUrl } from 'fragmentarium/ui/FragmentLink'

export default class FragmentCrumb implements Crumb {
  readonly text: string

  constructor(text: string) {
    this.text = text
  }

  get link(): string {
    return createFragmentUrl(this.text)
  }
}
