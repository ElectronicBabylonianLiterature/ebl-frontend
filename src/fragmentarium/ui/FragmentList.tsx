import React from 'react'
import { Table } from 'react-bootstrap'
import _ from 'lodash'
import FragmentLink from 'fragmentarium/ui/FragmentLink'
import './FragmentList.css'
import { FragmentInfo } from 'fragmentarium/domain/fragment'

export type Columns = Record<
  string,
  string | ((fragmentInfo: FragmentInfo) => React.ReactNode)
>

function FragmentList({
  fragments,
  columns,
}: {
  fragments: readonly FragmentInfo[]
  columns: Columns
}): JSX.Element {
  return (
    <Table responsive>
      <thead>
        <tr>
          <th>Number</th>
          {_.keys(columns).map((heading, index) => (
            <th className={`FragmentList__${heading}`} key={index}>
              {heading}
            </th>
          ))}
        </tr>
      </thead>
      <tbody>
        {fragments.map((fragment) => (
          <tr key={fragment.number}>
            <td>
              <FragmentLink number={fragment.number}>
                {fragment.number}
              </FragmentLink>
            </td>
            {_.values(columns).map((property, index) => (
              <td key={index}>
                {_.isFunction(property)
                  ? property(fragment)
                  : property
                      .split('.')
                      .reduce((object, index) => object[index], fragment)}
              </td>
            ))}
          </tr>
        ))}
        {_.isEmpty(fragments) && (
          <tr>
            <td colSpan={4}>No fragments found.</td>
          </tr>
        )}
      </tbody>
    </Table>
  )
}

export default FragmentList
