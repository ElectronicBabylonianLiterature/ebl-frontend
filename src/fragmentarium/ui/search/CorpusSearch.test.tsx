import React from 'react'
import { Router } from 'react-router-dom'
import { render, screen, waitFor } from '@testing-library/react'
import Promise from 'bluebird'
import Bluebird from 'bluebird'
import SessionContext from 'auth/SessionContext'
import MemorySession, { Session } from 'auth/Session'
import TextService from 'corpus/application/TextService'
import { fromMatchingLineDto } from 'corpus/application/dtos'
import WordService from 'dictionary/application/WordService'
import textLineFixture from 'test-support/lines/text-line'
import { stageToAbbreviation } from 'corpus/domain/period'
import { DictionaryContext } from 'dictionary/ui/dictionary-context'
import CorpusTransliterationSearch from 'fragmentarium/ui/search/ChapterInfoResults'
import { waitForSpinnerToBeRemoved } from 'test-support/waitForSpinnerToBeRemoved'
import userEvent from '@testing-library/user-event'
import { createMemoryHistory, MemoryHistory } from 'history'
import ChapterInfo from 'corpus/domain/ChapterInfo'

jest.mock('fragmentarium/application/FragmentSearchService')
jest.mock('corpus/application/TextService')
jest.mock('dictionary/application/WordService')

let history: MemoryHistory

const wordService = new (WordService as jest.Mock<WordService>)()
const session: Session = new MemorySession(['read:fragments'])
let container: HTMLElement

const corpusResult: ChapterInfo = {
  id: {
    textId: {
      genre: 'L',
      category: 1,
      index: 2,
    },
    stage: 'Old Babylonian',
    name: 'My Chapter',
  },
  siglums: { '1': 'NinSchb' },
  textName: 'Text Name Test',
  matchingLines: [
    fromMatchingLineDto({
      number: '1',
      isBeginningOfSection: false,
      isSecondLineOfParallelism: false,
      translation: [
        {
          language: 'de',
          extent: null,
          parts: [
            {
              text: 'Test text',
              type: 'StringPart',
            },
          ],
          content: [],
        },
      ],
      variants: [
        {
          reconstruction: '%n ra',
          manuscripts: [
            {
              manuscriptId: 1,
              labels: ['o', 'iii'],
              number: 'a+1',
              atf: 'ra',
              omittedWords: [],
            },
          ],
        },
      ],
    }),
  ],
  matchingColophonLines: {
    '1': [textLineFixture],
  },
}
const textService = new (TextService as jest.Mock<jest.Mocked<TextService>>)()

async function renderCorpusSearch(transliteration = 'test') {
  history = createMemoryHistory()
  jest.spyOn(history, 'push')
  const container = render(
    <Router history={history}>
      <DictionaryContext.Provider value={wordService}>
        <SessionContext.Provider value={session}>
          <CorpusTransliterationSearch
            paginationIndex={0}
            transliteration={transliteration}
            textService={textService}
          />
        </SessionContext.Provider>
      </DictionaryContext.Provider>
    </Router>
  ).container
  await waitForSpinnerToBeRemoved(screen)
  return container
}

describe('Corpus results', () => {
  beforeEach(async () => {
    textService.searchTransliteration.mockReturnValue(
      Promise.resolve({ chapterInfos: [corpusResult], totalCount: 1 })
    )
    container = await renderCorpusSearch()
  })

  it('Match Snapshot', () => {
    expect(container).toMatchSnapshot()
  })

  it('Name links to chapter', async () => {
    expect(await screen.findByText('1.')).toHaveAttribute(
      'href',
      `/corpus/${corpusResult.id.textId.genre}/${
        corpusResult.id.textId.category
      }/${corpusResult.id.textId.index}/${stageToAbbreviation(
        corpusResult.id.stage
      )}/${corpusResult.id.name}`
    )
  })
})

const transliteration = 'LI₂₃ ši₂-ṣa-pel₃-ṭa₃'
const TOTAL_COUNT = 2
describe('test scrolling through pagination', () => {
  beforeEach(async () => {
    textService.searchTransliteration
      .mockReturnValueOnce(
        Bluebird.resolve({
          chapterInfos: [corpusResult],
          totalCount: TOTAL_COUNT,
        })
      )
      .mockReturnValueOnce(
        Bluebird.resolve({
          chapterInfos: [
            {
              ...corpusResult,
              textName: ' Test Text Name 1',
            },
          ],
          totalCount: TOTAL_COUNT,
        })
      )
    await renderCorpusSearch(transliteration)
  })
  it('Inital Render', async () => {
    await waitFor(() =>
      expect(textService.searchTransliteration).toBeCalledWith(
        transliteration,
        0
      )
    )
    await screen.findByText(/Literature > Text Name Test/)
  })
  it('Next Page', async () => {
    userEvent.click(screen.getByText('2'))
    await waitFor(() =>
      expect(history.push).toHaveBeenCalledWith({
        search: 'paginationIndexCorpus=1',
      })
    )
    await waitFor(() =>
      expect(textService.searchTransliteration).toBeCalledWith(
        transliteration,
        1
      )
    )
    await screen.findByText(
      /Literature > Test Text Name 1 > Old Babylonian My Chapter/
    )
    expect(
      screen.queryByText(
        /Literature > Text Name Test > Old Babylonian My Chapter/
      )
    ).not.toBeInTheDocument()
  })
})
