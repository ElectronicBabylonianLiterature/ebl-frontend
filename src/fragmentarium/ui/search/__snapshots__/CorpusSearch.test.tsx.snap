// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Corpus results Match Snapshot 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="mt-2 col offset-5"
    >
      <ul
        class="pagination"
      >
        <li
          class="page-item active"
        >
          <span
            class="page-link"
          >
            1
            <span
              class="sr-only"
            >
              (current)
            </span>
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <hr />
      <div
        class="justify-content-center row"
      >
        <h5
          class="text-secondary"
        >
          Literature
           &gt; Text Name Test
           &gt; 
          Old Babylonian My Chapter
        </h5>
      </div>
      <div
        class="row"
      >
        <span
          class="lead"
        >
          <a
            href="/corpus/L/1/2/OB/My Chapter"
          >
            1
            .
          </a>
        </span>
      </div>
      <div
        class="row"
      >
        <div
          class="row"
        >
          <div
            class="ml-4 col"
          >
            <div
              class="lead"
            >
              <span>
                Test text
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="ml-5 pl-1 col"
        >
          NinSchb
           
          o iii
           
          a+1
          .
           
        </div>
      </div>
      <br />
      <div
        class="row"
      >
        NinSchb
         
        1
         
        <span
          class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
        >
           
        </span>
        <span
          class="Transliteration__Word Transliteration__Word--AKKADIAN"
        >
          <span
            class=""
          >
            <span
              class="Transliteration__Reading"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__ValueToken"
                >
                  <span
                    class=""
                  >
                    érin
                  </span>
                </span>
                <sup
                  class="Transliteration__modifier"
                />
                <sup
                  class="Transliteration__flag"
                />
              </span>
            </span>
            <span
              class="Transliteration__Joiner"
            >
              <span
                class=""
              >
                -
              </span>
            </span>
            <span
              class="Transliteration__BrokenAway"
            >
              <span
                class=""
              >
                [
              </span>
            </span>
            <span
              class="Transliteration__PerhapsBrokenAway"
            >
              <span
                class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY"
              >
                (
              </span>
            </span>
            <span
              class="Transliteration__Logogram"
            >
              <span
                class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
              >
                <span
                  class="Transliteration__ValueToken"
                >
                  <span
                    class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                  >
                    ŠÁ
                  </span>
                </span>
                <sup
                  class="Transliteration__modifier"
                />
                <sup
                  class="Transliteration__flag"
                />
              </span>
            </span>
            <span
              class="Transliteration__Joiner"
            >
              <span
                class="Transliteration__Joiner--BROKEN_AWAY Transliteration__Joiner--PERHAPS_BROKEN_AWAY"
              >
                -
              </span>
            </span>
            <span
              class="Transliteration__Variant"
            >
              <span
                class="Transliteration__Reading"
              >
                <span
                  class="Transliteration__Reading--BROKEN_AWAY Transliteration__Reading--PERHAPS_BROKEN_AWAY"
                >
                  <span
                    class="Transliteration__ValueToken"
                  >
                    <span
                      class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                    >
                      šà
                    </span>
                  </span>
                  <sup
                    class="Transliteration__modifier"
                  />
                  <sup
                    class="Transliteration__flag"
                  />
                </span>
              </span>
              /
              <span
                class="Transliteration__Logogram"
              >
                <span
                  class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                >
                  <span
                    class="Transliteration__ValueToken"
                  >
                    <span
                      class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                    >
                      GÌRI
                    </span>
                  </span>
                  <sup
                    class="Transliteration__modifier"
                  />
                  <sup
                    class="Transliteration__flag"
                  />
                </span>
              </span>
            </span>
            <span
              class="Transliteration__PerhapsBrokenAway"
            >
              <span
                class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY Transliteration__PerhapsBrokenAway--PERHAPS_BROKEN_AWAY"
              >
                )
              </span>
            </span>
          </span>
        </span>
        <span
          class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
        >
           
        </span>
        <span
          class="Transliteration__Word Transliteration__Word--AKKADIAN"
        >
          <span
            class=""
          >
            <span
              class="Transliteration__Reading"
            >
              <span
                class=""
              >
                <span
                  class="Transliteration__ValueToken"
                >
                  <span
                    class=""
                  >
                    kur
                  </span>
                </span>
                <sup
                  class="Transliteration__modifier"
                />
                <sup
                  class="Transliteration__flag"
                />
              </span>
            </span>
          </span>
        </span>
        <span
          class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
        >
           
        </span>
        <span
          class="Transliteration__Word Transliteration__Word--AKKADIAN"
        >
          <span
            class="Transliteration__Word--BROKEN_AWAY"
          >
            <span
              class="Transliteration__Reading"
            >
              <span
                class="Transliteration__Reading--BROKEN_AWAY"
              >
                <span
                  class="Transliteration__ValueToken"
                >
                  <span
                    class="Transliteration__ValueToken--BROKEN_AWAY"
                  >
                    saḫ
                  </span>
                </span>
                <sup
                  class="Transliteration__modifier"
                />
                <sup
                  class="Transliteration__flag"
                />
              </span>
            </span>
            <span
              class="Transliteration__Joiner"
            >
              <span
                class="Transliteration__Joiner--BROKEN_AWAY"
              >
                -
              </span>
            </span>
            <span
              class="Transliteration__Logogram"
            >
              <span
                class="Transliteration__Logogram--BROKEN_AWAY"
              >
                <span
                  class="Transliteration__ValueToken"
                >
                  <span
                    class="Transliteration__ValueToken--BROKEN_AWAY"
                  >
                    SAH
                  </span>
                </span>
                <sup
                  class="Transliteration__modifier"
                />
                <sup
                  class="Transliteration__flag"
                />
              </span>
            </span>
          </span>
        </span>
        <span
          class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
        >
           
        </span>
        <span
          class="Transliteration__UnknownNumberOfSigns Transliteration__UnknownNumberOfSigns--AKKADIAN"
        >
          …
        </span>
        <span
          class="Transliteration__BrokenAway Transliteration__BrokenAway--AKKADIAN"
        >
          <span
            class="Transliteration__BrokenAway--BROKEN_AWAY"
          >
            ]
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;
